### PR TITLE
Add AuthorizationBuilder

### PR DESCRIPTION
**Pull request by HaoK**
AuthZ part of https://github.com/dotnet/aspnetcore/issues/42235

I didn't update everything, but updated enough of the tests to feel confident that this does reduce lines of code even just in our tests, and that the basics are working, will finish the rest of this assuming this is something that we want to take for real.

Before
```
      var authorizationService = BuildAuthorizationService(services =>
        {
            services.AddAuthorization(options =>
            {
                options.AddPolicy("Basic", policy => policy.RequireClaim("Permission", "CanViewPage"));
            });
        });
```

After
```
            services.AddAuthorization().AddPolicy("Basic", policy => policy.RequireClaim("Permission", "CanViewPage")));```
```

Note: this does change the return value for AddAuthorization* methods to return the new AuthorizationBuilderServiceCollection that wraps the service collection

cc @DamianEdwards @davidfowl @captainsafia @adityamandaleeka

### CODE REVIEW

File: src/Security/Authorization/Core/src/AuthorizationServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -17,8 +17,8 @@ public static class AuthorizationServiceCollectionExtensions
     /// Adds authorization services to the specified <see cref="IServiceCollection" />.
     /// </summary>
     /// <param name="services">The <see cref="IServiceCollection" /> to add services to.</param>
-    /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
-    public static IServiceCollection AddAuthorizationCore(this IServiceCollection services)
+    /// <returns>The <see cref="AuthorizationBuilderServiceCollection"/> so that additional calls can be chained.</returns>
+    public static AuthorizationBuilderServiceCollection AddAuthorizationCore(this IServiceCollection services)
```
Comment (davidfowl) [inline src/Security/Authorization/Core/src/AuthorizationServiceCollectionExtensions.cs:21]: This is a breaking change.

Comment (HaoK) [reply src/Security/Authorization/Core/src/AuthorizationServiceCollectionExtensions.cs:21]: Is this one we can get approved, functionally it shouldn't affect most people right?

**Comment by HaoK** [comment]
Updated with new proposed AddAuthorizationBuilder name

**Comment by HaoK** [comment]
@DamianEdwards is this something you want me to get in for preview6, or can this wait for preview7?  The PR Is basically ready, since I don't have to update all tests to use the new API

**Comment by DamianEdwards** [comment]
It would still need an API review and we should likely do this change at the same time as introducing the new `WebApplicationBuilder.Authorization` property. Code complete is tomorrow and we still have quite a lot in preview.6 to land in an SDK build so might be wise to push this to preview.7

**Comment by HaoK** [comment]
@captainsafia do you want make the WAB/Authorization side of things in my branch or do you want me to do it (for preview 7)?

**Comment by davidfowl** [comment]
Lets put this up for API review (write an API proposal based on this PR in the issue)

**Review by Tratcher — APPROVED** [review]

File: src/Security/Authorization/Core/src/AuthorizationBuilder.cs [hunk 2]
```diff
@@ -29,7 +29,7 @@ public AuthorizationBuilder(IServiceCollection services)
     /// Defaults to true.
     /// </summary>
     /// <returns>The builder.</returns>
-    public virtual AuthorizationBuilder SetInvokeHandlersAfterFailure(bool invoke)
+    public virtual AuthorizationBuilder InvokeHandlersAfterFailure(bool invoke)
```
Comment (davidfowl) [inline src/Security/Authorization/Core/src/AuthorizationBuilder.cs:32]: Set* seems better tbh

Comment (HaoK) [reply src/Security/Authorization/Core/src/AuthorizationBuilder.cs:32]: Okay, going back to the original SetInvokeHandlersAfterFailure

**Pull request merged**